hw/src/ledger.rs

```diff
@@ -0,0 +1,364 @@
+	UserCancel,
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		match *self {
+			Error::Protocol(ref s) => write!(f, "Ledget protocol error: {}", s),
+			Error::Usb(ref e) => write!(f, "USB communication error: {}", e),
+			Error::KeyNotFound => write!(f, "Key not found"),
+			Error::UserCancel => write!(f, "Operation has been cancelled"),
+		}
+	}
+}
+
+impl From<hidapi::HidError> for Error {
+	fn from(err: hidapi::HidError) -> Error {
+		Error::Usb(err)
+	}
+}
+
+/// Ledger device manager.
+pub struct Manager {
+	usb: hidapi::HidApi,
+	devices: Vec<Device>,
+	key_path: KeyPath,
+}
+
+#[derive(Debug)]
+struct Device {
+	path: String,
+	info: WalletInfo,
+}
+
+impl Manager {
+	/// Create a new instance.
+	pub fn new() -> Result<Manager, Error> {
+		let manager = Manager {
+			usb: hidapi::HidApi::new()?,
+			devices: Vec::new(),
+			key_path: KeyPath::Ethereum,
+		};
+		Ok(manager)
+	}
+
+	/// Re-populate device list. Only those devices that have Ethereum app open will be added.
+	pub fn update_devices(&mut self) -> Result<usize, Error> {
+		self.usb.refresh_devices();
+		let devices = self.usb.devices();
+		let mut new_devices = Vec::new();
+		let mut num_new_devices = 0;
+		for device in devices {
+			trace!("Checking device: {:?}", device);
+			if device.vendor_id != LEDGER_VID || !LEDGER_PIDS.contains(&device.product_id) {
+				continue;
+			}
+			match self.read_device_info(&device) {
+				Ok(info) => {
+					debug!("Found device: {:?}", info);
+					if !self.devices.iter().any(|d| d.path == info.path) {
+						num_new_devices += 1;
+					}
+					new_devices.push(info);
+
+				},
+				Err(e) => debug!("Error reading device info: {}", e),
+			};
+		}
+		self.devices = new_devices;
+		Ok(num_new_devices)
+	}
+
+	/// Select key derivation path for a known chain.
+	pub fn set_key_path(&mut self, key_path: KeyPath) {
+		self.key_path = key_path;
+	}
+
+	fn read_device_info(&self, dev_info: &hidapi::HidDeviceInfo) -> Result<Device, Error> {
+		let mut handle = self.open_path(&dev_info.path)?;
+		let address = Self::read_wallet_address(&mut handle, self.key_path)?;
+		let manufacturer = dev_info.manufacturer_string.clone().unwrap_or("Unknown".to_owned());
+		let name = dev_info.product_string.clone().unwrap_or("Unknown".to_owned());
+		let serial = dev_info.serial_number.clone().unwrap_or("Unknown".to_owned());
+		Ok(Device {
+			path: dev_info.path.clone(),
+			info: WalletInfo {
+				name: name,
+				manufacturer: manufacturer,
+				serial: serial,
+				address: address,
+			},
+		})
+	}
+
+	fn read_wallet_address(handle: &hidapi::HidDevice, key_path: KeyPath) -> Result<Address, Error> {
+		let ver = Self::send_apdu(handle, commands::GET_APP_CONFIGURATION, 0, 0, &[])?;
+		if ver.len() != 4 {
+			return Err(Error::Protocol("Version packet size mismatch"));
+		}
+
+		let (major, minor, patch) = (ver[1], ver[2], ver[3]);
+		if major <= 1 && minor == 0 && patch < 3 {
```

Comment:
Is this condition correct? Should it mean that anything greater then 1.0.3 is supported?
If so then it needs to be:
```rust
if major < 1 || (major == 1 && minor == 0 && patch < 3) 
```
Currently version `0.x.0` (`x > 0`) will be treated as correct, no?